ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -336,97 +337,81 @@ object KeyValueCommitting {
-
-        // 1. Verify that the submitter can view the contract.
-        _ <- if (locallyDisclosedTo.contains(submitter) || divulgedTo.contains(submitter))
-          Some(())
-        else {
-          tracelog(s"lookupContract($coid): Contract not visible to submitter $submitter")
-          None
-        }
-
-        // 2. Verify that the contract is active
-        _ <- if (contractState.hasActiveAt && txLet >= parseTimestamp(contractState.getActiveAt))
-          Some(())
-        else {
-          val activeAtStr =
-            if (contractState.hasActiveAt)
-              parseTimestamp(contractState.getActiveAt).toString
-            else "<activeAt missing>"
-          tracelog(
-            s"lookupContract($coid): Contract not active (let=$txLet, activeAt=$activeAtStr).")
-          None
-        }
-
-        // Finally lookup the log entry containing the create node and the contract instance.
-        entry = inputLogEntries
-          .getOrElse(eid, throw Err.MissingInputLogEntry(eid))
-        coinst <- lookupContractInstanceFromLogEntry(eid, entry, nid)
-      } yield (coinst)
+    def contractVisibleToSubmitter(contractState: DamlContractState): Boolean = {
+      val locallyDisclosedTo = contractState.getLocallyDisclosedToList.asScala.toSet
+      val divulgedTo = contractState.getDivulgedToList.asScala.toSet
+      locallyDisclosedTo.contains(submitter) || divulgedTo.contains(submitter)
     }
 
-    // Helper to lookup package from the state. The package contents
-    // are stored in the [[DamlLogEntry]], which we find by looking up
-    // the DAML state entry at `DamlStateKey(packageId = pkgId)`.
-    def lookupPackage(pkgId: PackageId) = {
-      val stateKey = DamlStateKey.newBuilder.setPackageId(pkgId).build
-      for {
-        value <- inputState
-          .get(stateKey)
-          .flatten
-          .orElse {
-            throw Err.MissingInputState(stateKey)
-          }
-        pkg <- value.getValueCase match {
-          case DamlStateValue.ValueCase.ARCHIVE =>
-            // NOTE(JM): Engine only looks up packages once, compiles and caches,
-            // provided that the engine instance is persisted.
-            Some(Decode.decodeArchive(value.getArchive)._2)
-          case _ =>
-            tracelog(s"lookupPackage($pkgId): value not a DAML-LF archive!")
-            None
-        }
-      } yield pkg
-    }
+    // Pull all keys from referenced contracts. We require this for 'fetchByKey' calls
+    // which are not evidenced in the transaction itself and hence the contract key state is
+    // not included in the inputs.
+    lazy val knownKeys: Map[GlobalKey, AbsoluteContractId] =
+      inputState.toList.collect {
+        case (key, Some(value))
+            if value.hasContractState
+              && value.getContractState.hasContractKey
+              && contractVisibleToSubmitter(value.getContractState) =>
+          Conversions.decodeContractKey(value.getContractState.getContractKey) ->
```

Comment:
I'm not familiar with this code, but I suspect that if a transaction archives a contract with a key and then creates another contract with the same key, then the creation may overwrite the key entry for the actual input contract in this map.

Reply:
Shouldn't be a problem. This code wouldn't get invoked if the contract is archived and new one created with the same key since the interpreter already catches this.

Reply:
But that's a perfectly legitimate use case. First archive the contract with the key and then create a new one with the same key, all in the same transaction. DAMLe is supposed to produce such a transaction. I'm worried that this code visits the create node and later fetch nodes and this overwrites the data from the lookupByKey/exercise that archives the input contract.